File: itmd-521/labs/module-04/py/module-04.py
import sys 

from pyspark.sql import SparkSession
from pyspark.sql.functions import *
from pyspark.sql.types import *

if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit(-1)
    

    spark = (SparkSession.builder.appName("Divvy Print Schema").getOrCreate())

    divvy_file = sys.argv[1] 
    
    df_infer = (spark.read.format("csv").option("header","true").option("inferSchema","true").load(divvy_file))
    
    df_infer.printSchema()
    print(f"Row count with inferred schema: {df_infer.count()}")

    programmatic_schema = StructType([
    StructField("trip_id", IntegerType(), False),
    StructField("starttime", TimestampType(), True),
    StructField("stoptime", TimestampType(), True),
    StructField("bikeid", IntegerType(), True),
    StructField("tripduration", IntegerType(), True),
    StructField("from_station_id", IntegerType(), True),
    StructField("from_station_name", StringType(), True),
    StructField("to_station_id", IntegerType(), True),
    StructField("to_station_name", StringType(), True),
    StructField("usertype", StringType(), True),
    StructField("gender", StringType(), True),
    StructField("birthyear", IntegerType(), True)])
   
    df_programmatic = spark.read.csv(divvy_file, schema=programmatic_schema, header=True)
    
    df_programmatic.printSchema()
    print(f"Row count with programmatic schema: {df_programmatic.count()}")

    ddl_schema = """
    trip_id INT,
    starttime TIMESTAMP,
    stoptime TIMESTAMP,
    bikeid INT,
    tripduration INT,
    from_station_id INT,
    from_station_name STRING,
    to_station_id INT,
    to_station_name STRING,
    usertype STRING,
    gender STRING,
    birthyear INT 
    """
 
    df_ddl = spark.read.csv(divvy_file, schema=StructType.fromDDL(ddl_schema), header=True)

    df_ddl.printSchema()
    print(f"Row count with DDL schema: {df_ddl.count()}")

    spark.stop()

File: itmd-521/labs/module-04/module-04.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._

object DivvyPrintSchema {

  def main(args: Array[String]): Unit = {
    if (args.length != 1) {
      System.exit(-1)
    }

    // Initialize SparkSession
    val spark = SparkSession.builder
      .appName("Divvy Print Schema")
      .getOrCreate()

    val divvyFile = args(0)

    // Reading CSV with inferred schema
    val dfInferredSchema = spark.read
      .format("csv")
      .option("header", "true")
      .option("inferSchema", "true")
      .load(divvyFile)

    // Print inferred schema and count
    dfInferredSchema.printSchema()
    println(s"Row count with inferred schema: ${dfInferredSchema.count()}")

    // Define programmatic schema
    val programmaticSchema = StructType(Array(
      StructField("trip_id", IntegerType, false),
      StructField("starttime", TimestampType, true),
      StructField("stoptime", TimestampType, true),
      StructField("bikeid", IntegerType, true),
      StructField("tripduration", IntegerType, true),
      StructField("from_station_id", IntegerType, true),
      StructField("from_station_name", StringType, true),
      StructField("to_station_id", IntegerType, true),
      StructField("to_station_name", StringType, true),
      StructField("usertype", StringType, true),
      StructField("gender", StringType, true),
      StructField("birthyear", IntegerType, true)
    ))

    // Reading CSV with programmatic schema
    val dfProgrammatic = spark.read
      .format("csv")
      .schema(programmaticSchema)
      .option("header", "true")
      .load(divvyFile)

    // Print programmatic schema and count
    dfProgrammatic.printSchema()
    println(s"Row count with programmatic schema: ${dfProgrammatic.count()}")

    // Define DDL schema
    val ddlSchema = """
      trip_id INT,
      starttime TIMESTAMP,
      stoptime TIMESTAMP,
      bikeid INT,
      tripduration INT,
      from_station_id INT,
      from_station_name STRING,
      to_station_id INT,
      to_station_name STRING,
      usertype STRING,
      gender STRING,
      birthyear INT
    """

    // Reading CSV with DDL schema
    val dfDDL = spark.read
      .format("csv")
      .schema(ddlSchema)
      .option("header", "true")
      .load(divvyFile)

    // Print DDL schema and count
    dfDDL.printSchema()
    println(s"Row count with DDL schema: ${dfDDL.count()}")

    // Stop the Spark session
    spark.stop()
  }
}
